src/test/java/com/google/api/gax/grpc/UnaryCallableTest.java

```diff
@@ -602,6 +621,57 @@ public void bundling() throws Exception {
     }
   }
 
+  @Test
+  public void bundlingWithFlowControl() throws Exception {
+    BundlingSettings bundlingSettings =
+        BundlingSettings.newBuilder()
+            .setDelayThreshold(Duration.standardSeconds(1))
+            .setElementCountThreshold(4L)
+            .setFlowControlSettings(
+                FlowControlSettings.newBuilder()
+                    .setLimitExceededBehavior(LimitExceededBehavior.Block)
+                    .setMaxOutstandingElementCount(10)
+                    .setMaxOutstandingRequestBytes(10)
+                    .build())
+            .build();
+    TrackedFlowController trackedFlowController =
+        new TrackedFlowController(bundlingSettings.getFlowControlSettings());
+    BundlerFactory<LabeledIntList, List<Integer>> bundlerFactory =
+        new BundlerFactory<>(
+            SQUARER_BUNDLING_DESC, bundlingSettings, bundlingExecutor, trackedFlowController);
+
+    Truth.assertThat(trackedFlowController.getElementsReserved()).isEqualTo(0);
+    Truth.assertThat(trackedFlowController.getElementsReleased()).isEqualTo(0);
+    Truth.assertThat(trackedFlowController.getBytesReserved()).isEqualTo(0);
+    Truth.assertThat(trackedFlowController.getBytesReleased()).isEqualTo(0);
+    Truth.assertThat(trackedFlowController.getCallsToReserve()).isEqualTo(0);
+    Truth.assertThat(trackedFlowController.getCallsToRelease()).isEqualTo(0);
+
+    try {
+      UnaryCallable<LabeledIntList, List<Integer>> callable =
+          UnaryCallable.<LabeledIntList, List<Integer>>create(callLabeledIntSquarer)
+              .bundling(SQUARER_BUNDLING_DESC, bundlerFactory);
+      ApiFuture<List<Integer>> f1 = callable.futureCall(new LabeledIntList("one", 1, 2));
+      ApiFuture<List<Integer>> f2 = callable.futureCall(new LabeledIntList("one", 3, 4));
+      Truth.assertThat(f1.get()).isEqualTo(Arrays.asList(1, 4));
+      Truth.assertThat(f2.get()).isEqualTo(Arrays.asList(9, 16));
+    } finally {
+      bundlerFactory.close();
+    }
+
+    // Give time for the executor to complete tasks to release resources
+    Thread.sleep(100);
```

Comment:
Have modified `pushCurrentBundle` to return a future. Also made package-private as we do not want to be constrained to keep this in the surface.